Update to draft to address issues (#22)
ietf-wg-jose/draft-ietf-jose-hpke-encrypt: draft-ietf-jose-hpke-encrypt.md

  *  HPKE JWE Integrated Encryption, where HPKE is used to encrypt the plaintext.
  *  HPKE JWE Key Encryption, where HPKE is used to encrypt a content encryption key (CEK) and the CEK is subsequently used to encrypt the plaintext.

When "alg" is a JOSE-HPKE algorithm:

  * If "enc" is "dir", HPKE JWE Integrated Encryption is used.

>   * If "enc" is an AEAD algorithm, The recipient Key Managment mode is Key Encryption.

Reviewer: ```suggestion
  * If "enc" is an AEAD algorithm, the recipient Key Managment mode is Key Encryption.
```

> The HPKE KEM, KDF and AEAD used depend on the JOSE-HPKE algorithm used.

Reviewer: ```suggestion
The HPKE KEM, KDF, and AEAD used depend on the JOSE-HPKE algorithm used.
```

- The "encrypted_key" MUST be the base64url encoded encapsulated key as defined in Section 5.1.1 of {{RFC9180}}.
- The "iv", "tag" and "aad" members MUST NOT be present.
- The "ciphertext" MUST be the base64url encoded ciphertext as defined in Section 5.2 of {{RFC9180}}.

> - The HPKE Setup info parameter MAY be used, and its values are not constrained by this specification.

Reviewer: This doesn't say how the HPKE Setup Info parameter value is provided.

Author: As per our discussion in the call, I updated the text to align with COSE. 
The HPKE Setup info parameter MUST be set to an empty string.

> - The HPKE AEAD AAD MUST be set to the "JWE Additional Authenticated Data encryption parameter" defined in Step 14 of Section 5.1 of {{RFC7516}}.

Reviewer: ```suggestion
- The HPKE AEAD AAD MUST be set to the "JWE Additional Authenticated Data encryption parameter", as defined in Step 14 of Section 5.1 of {{RFC7516}}.
```

> If the "zip" header parameter is present, the resulting plaintext is uncompressed using the algorithm specified and the result is the
> raw message plaintext.
> 

Reviewer: ```suggestion
```

Talking about zip is unnecessary here because JWE specifies this behavior, which applies to all algorithms.

Author: The text is only specified for the new JWE Integrated Encryption mode.

Reviewer 2: A reference would be better than repeating the text... some thing like:

Not that compression is possible with both integrated encryption and key encryption, see section .... of rfc ...

Author: I don't see a need to re-iterate that compression is possible with key encryption but I agree to avoid repeating the text.

> raw message plaintext. Otherwise the resulting plaintext is the raw message plaintext.

Reviewer 2: ```suggestion
raw message plaintext.
```

Reviewer 2: not critical, but i dont think we need this sentence

When decrypting, the checks in {{RFC7516}} section 5.2, steps 1 through 5 MUST be performed.

> 

Author: The text is only specified for the new JWE Integrated Encryption mode.

> 

Reviewer 2: Probably we should add a comment that "JWE Encrypted Key" in step 2 is encapsulated key.

> 

Author: Thanks, fixed.

- The "iv", "tag" JWE members MUST be present.
- The "aad" JWE member MAY be present.
- The "ciphertext" MUST be the base64url encoded ciphertext as described in Step 19 in Section 5.1 of {{RFC7516}}.

> - The HPKE Setup info parameter MUST be set to an empty string.

Reviewer 2: ```suggestion
- The HPKE Setup info parameter SHOULD be set to an empty string. See RFC9180 Section 8.1 for details on how to provide auxiliary authenticated application information.
```

Author: The COSE-HPKE draft states that the setup info is set to an empty string. We should ensure that both drafts are aligned. In the last meeting, Mike was of the opinion that a new JOSE parameter would need to be added if auxiliary authenticated application information is required to be provided to HPKE setup info.

Reviewer: I agree with making this consistent with COSE HPKE by making the parameter empty.  Let's deal with possibly relaxing this in a separate issue, if needed, but not in this PR.

> - The HPKE Setup info parameter MAY be used, and its values are not constrained by this specification.

Reviewer: This doesn't say how the HPKE Setup Info parameter value is provided.

HPKE relies on a source of randomness to be available on the device.
In Key Agreement with Key Wrapping mode, CEK has to be randomly generated and it MUST be ensured that the guidelines in {{RFC8937}} for random number generations are followed.

## HPKE authentication

Authenticated HPKE modes MUST NOT be used for Key Encryption, as they do not authenticate the message. These modes validate the sender's identity by ensuring that the shared secret originates from the sender's static public key, a pre-shared key, or a combination of both. However, this authentication is limited to the key exchange process and does not cover the message encrypted with the CEK. For example, a malicious recipient could decrypt the CEK, modify the decrypted message, and re-encrypt it using the same CEK. Other recipients, unaware of the tampering, might mistakenly trust the altered message as if it came from the authenticated sender.

## Key Management

> A single key MUST NOT be used in both sender and recipient roles. Avoiding the use of the same key for both sender and recipient roles ensures clear cryptographic boundaries and minimizes unintended interactions.

Reviewer 2: This prevents encrypting messages to yourself, which part of the HPKE specification makes this requirement?

Author: Encrypting messages to yourself is typically done using symmetric algorithms, For instance, in case of PQC KEM, it looks complex to encrypt the symmetric key to yourself using encap and then later using decap to get the symmetric key. I don't see any benefit for the additional compute.

Reviewer: I'm not sure why the good guidance above is being deleted.

Author: As per our discussion in the call, reverted the text, it will take up in a separate PR.

> A single key MUST NOT be used with both JOSE-HPKE and other algorithms as this might enable cross-protocol attacks.

Reviewer 2: ```suggestion
A single recipient or sender key MUST NOT be used with both JOSE-HPKE and other algorithms as this might enable cross-protocol attacks.
```

> A single recipient or sender key MUST NOT be used with both JOSE-HPKE and other algorithms as this might enable cross-protocol attacks.

Reviewer: Not reusing a key across multiple algorithms is normal cryptographic hygiene, and not specific to HPKE.  Do we really need to say this here?

In my view, we should only be adding text about how to use HPKE with JWE.

Author: Explicitly stating this helps ensure that implementers are aware of and avoid potential cross-protocol attack surfaces.

Reviewer 2: Making this informative or and moving it to security considerations is perhaps acceptable middle ground.

Author: The above text is already in Security Considerations section, please check.

> A single key MAY be used with both Integrated Encryption and Key Encryption.

Reviewer: Why is this OK?  This seems like a form of key reuse.

Author: As discussed in the call, the text is reverted.

HPKE relies on a source of randomness to be available on the device.
In Key Agreement with Key Wrapping mode, CEK has to be randomly generated and it MUST be ensured that the guidelines in {{RFC8937}} for random number generations are followed.

> ## HPKE authentication

Reviewer 2: ```suggestion
## Authentication using an Asymmetric Key
```

> Authenticated HPKE modes MUST NOT be used for Key Encryption, as they do not authenticate the message. These modes validate the sender's identity by ensuring that the shared secret originates from the sender's static public key, a pre-shared key, or a combination of both. However, this authentication is limited to the key exchange process and does not cover the message encrypted with the CEK. For example, a malicious recipient could decrypt the CEK, modify the decrypted message, and re-encrypt it using the same CEK. Other recipients, unaware of the tampering, might mistakenly trust the altered message as if it came from the authenticated sender.

Reviewer 2: ```suggestion
Implementers are cautioned to note that the use of authenticated KEMs has different meaning when considering integrated encryption and key encryption.
In integrated encryption the KEM operations secure the message plaintext, whereas with key encryption, the KEM operations secure the content encryption key.
For this reason, the use of authenticated KEMs with key encryption is NOT RECOMMENDED, as it gives a false sense of security.
See RFC9180 Section 5.1.3 for details authentication using asymmetric keys.
```

## Key Management

Reusing a single KEM key across multiple algorithm combinations MUST be avoided to maintain cryptographic security.  Each key and its
associated algorithm suite, comprising the KEM, KDF, and AEAD, should be managed independently.  This separation prevents unintended
interactions or vulnerabilities between suites, ensuring the integrity and security guarantees of each algorithm suite are

> preserved.  Additionally, the same key MUST NOT be used for both key wrapping and content encryption, as it may introduce security risks.

Reviewer 2: ```suggestion
preserved.  Additionally, the same key MUST NOT be used for both key encryption and integrated encryption, as it may introduce security risks.
```
